config/config.go

```diff
@@ -0,0 +1,118 @@
+package config
+
+import (
+	"io/ioutil"
+	"path"
+	"gopkg.in/yaml.v2"
+	"github.com/gruntwork-io/boilerplate/util"
+	"fmt"
+	"github.com/gruntwork-io/boilerplate/errors"
+)
+
+const BOILERPLATE_CONFIG_FILE = "boilerplate.yml"
+
+const OPT_TEMPLATE_FOLDER = "template-folder"
+const OPT_OUTPUT_FOLDER = "output-folder"
+const OPT_NON_INTERACTIVE = "non-interactive"
+const OPT_VAR = "var"
+const OPT_VAR_FILE = "var-file"
+
+// The command-line options for the boilerplate app
+type BoilerplateOptions struct {
+	TemplateFolder 	string
+	OutputFolder 	string
+	NonInteractive	bool
+	Vars		map[string]string
+}
+
+// Validate that the options have reasonable values and return an error if they don't
+func (options *BoilerplateOptions) Validate() error {
+	if options.TemplateFolder == "" {
+		return errors.WithStackTrace(TemplateFolderCannotBeEmpty)
+	}
+
+	if !util.PathExists(options.TemplateFolder) {
+		return errors.WithStackTrace(TemplateFolderDoesNotExist(options.TemplateFolder))
+	}
+
+	if options.OutputFolder == "" {
+		return errors.WithStackTrace(OutputFolderCannotBeEmpty)
+	}
+
+	return nil
+}
+
+// The contents of a boilerplate.yml config file
+type BoilerplateConfig struct {
+	Variables []Variable
+}
+
+// A single variable defined in a boilerplate.yml config file
+type Variable struct {
+	Name 	string
+	Prompt 	string
+	Default	string
+}
+
+// Return the default path for a boilerplate.yml config file in the given folder
+func BoilerPlateConfigPath(templateFolder string) string {
+	return path.Join(templateFolder, BOILERPLATE_CONFIG_FILE)
+}
+
+// Load the boilerplate.yml config contents for the folder specified in the given options
+func LoadBoilerPlateConfig(options *BoilerplateOptions) (*BoilerplateConfig, error) {
+	configPath := BoilerPlateConfigPath(options.TemplateFolder)
+
+	if util.PathExists(configPath) {
+		util.Logger.Printf("Loading boilerplate config from %s", configPath)
```

Comment:
Gah, goddamn VCs making people think that being "contrarian" is a good thing. 

The truth about logging is very simple: when you're debugging tricky production issues, you almost always wish you had either more info or less info. You want more info when errors are happening without leaving any obvious signs; you want less info when the errors are being drowned out by noise. The ability to control log levels is one way to do that.

It's immensely useful in all servers. Normally, you set it to a high level (INFO or higher) to keep noise low, but when requests start failing, or you have concurrency bugs, or something is slow, you crank it up to DEBUG or even TRACE and kick yourself in the ass for not having included more log output so you could see exactly what the code was doing.

It's immensely useful in CLI apps. Normally, you set it to a high level (INFO or higher) so it doesn't distract users with lots of useless noise. But when users report bugs, you ask them to set it to DEBUG or TRACE so you can figure out what the hell is going on, and kick yourself in the ass for not having included more log output so you could see exactly what the code was doing.

I have never, ever, ever wished I had included less logging or regretted using various log levels. I have very, very often wished I had included more logging, especially at the DEBUG or TRACE level.